File: HW05/polygonvect.cpp
#include "polygonvect.h"

namespace GtuObjects{
	
//Default constructor.
PolygonVect::PolygonVect() : Polygon() { /*Deliberately empty */}

//Sending a vector.
PolygonVect::PolygonVect(const std::vector < Point2D > &v) throw(EmptyVectorException) : Polygon() {
	if(v.size()==0){
		throw EmptyVectorException();
	}
	else{
		points = v;
	}
}

//Sending a point.
PolygonVect::PolygonVect(const Point2D &p) : Polygon() {
	points.push_back(p);
}

//Conversion constructor circle to polygonvect.
PolygonVect::PolygonVect(const Circle &circle) {

	Point2D p;

	for(int i=0;i<100;i++) {
  		p.setX((circle.getCx() + circle.getRadius() * cos(i * 2 * M_PI / 100)));	// X-axis for polygon.
  		p.setY((circle.getCy() + circle.getRadius() * sin(i * 2 * M_PI / 100)));	// Y-axis for polygon.
  		points.push_back(p);
	}

	color = circle.getColor();
}

//Conversion constructor rectangle to polygonvect.
PolygonVect::PolygonVect(const Rectangle &rect){

	Point2D p;

	p.setX(rect.getX());
	p.setY(rect.getY());
	points.push_back(p);
	p.setX(rect.getX());
	p.setY(rect.getY()+rect.getHeight());
	points.push_back(p);
	p.setX(rect.getX()+rect.getWidth());
	p.setY(rect.getY()+rect.getHeight());
	points.push_back(p);
	p.setX(rect.getX()+rect.getWidth());
	p.setY(rect.getY());
	points.push_back(p);
	
	color = rect.getColor();
}

//Conversion constructor triangle to polygonvect.
PolygonVect::PolygonVect(const Triangle &tri){
	Point2D p;

	p.setX(tri.get_x_axis_of_Point1());
	p.setY(tri.get_y_axis_of_Point1());
	points.push_back(p);
	p.setX(tri.get_x_axis_of_Point2());
	p.setY(tri.get_y_axis_of_Point2());
	points.push_back(p);
	p.setX(tri.get_x_axis_of_Point3());
	p.setY(tri.get_y_axis_of_Point3());
	points.push_back(p);

	color = tri.getColor();
}

//Comparison operators for polygonvect.
bool PolygonVect::operator==(const PolygonVect &other)const{

	return points==other.points;
}

bool PolygonVect::operator!=(const PolygonVect &other)const{
		
	return !(*this==other);
		
}


// Index operator overloading for PolygonVect
const Polygon::Point2D& PolygonVect::operator[](int index)const throw(OutOfBoundsException){
	
	if(index<0 || index >= points.size() ){
		throw OutOfBoundsException();	// Throwing exception.
	}
	else{
		return points[index];
	}
}


Polygon::Point2D& PolygonVect::operator[](int index) throw(OutOfBoundsException){

	if(index<0 || index >= points.size() ){
		throw OutOfBoundsException();	// Throwing exception.
	}
	else{
		return points[index];
	}
}

const PolygonVect operator+(const PolygonVect &p1,const PolygonVect &p2){

	PolygonVect temp;

	temp.points.reserve(p1.points.size() + p2.points.size());
	temp.points.insert(temp.points.end(),p1.points.begin(),p1.points.end());
	temp.points.insert(temp.points.end(),p2.points.begin(),p2.points.end());

	return temp;
}

//Virtual area function implementation.
double PolygonVect::Area() const {
	double area = 0;

	for(int i=0 ; i<points.size()-1 ; i++){
		area += (points[i].getX() * points[i+1].getY() - points[i].getY() * points[i+1].getX() );
	}
	area += (points[points.size()-1].getX()*points[0].getY() - points[points.size()-1].getY() * points[0].getX() );
	area = fabs(area)/2.0;

	
	return area;
	
}

//Virtual perimeter function implementation.
double PolygonVect::Perimeter() const {
	double perimeter = 0;


	for(int i=0 ; i<points.size()-1 ; i++){
		perimeter += sqrt(pow(points[i].getX()-points[i+1].getX(),2) + pow(points[i].getY() - points[i+1].getY(),2));
	}
	perimeter += sqrt(pow(points[points.size()-1].getX()-points[0].getX(),2) + pow(points[points.size()-1].getY() - points[0].getY(),2));

	return perimeter;

}

//Virtual helper draw function implementation.
std::ostream& PolygonVect::draw(std::ostream& output) const {

	output<< "<polygon points=\"";
	for(int i=0;i<points.size();i++){
		output << points[i].getX() << "," << points[i].getY() << " ";
	}
	
	output<< "\" style=\"fill:" << color << "\" />" << std::endl ;
	
	return output;

}

//İmplementation of post and prefix operator overloads.
Shape& PolygonVect::operator++() {

	for(int i=0 ; i<points.size() ; i++){
		points[i].setX(points[i].getX()+1);
		points[i].setY(points[i].getY()+1);
	}
	return *this;
}

Shape& PolygonVect::operator++(int){
	PolygonVect  *temp = new PolygonVect(*this);
	++*this;

	return *temp;

}

Shape& PolygonVect::operator--(){
	for(int i=0 ; i<points.size() ; i++){
		points[i].setX(points[i].getX()-1);
		points[i].setY(points[i].getY()-1);
	}
	return *this;
}

Shape& PolygonVect::operator--(int){
	PolygonVect  *temp = new PolygonVect(*this);
	--*this;

	return *temp;
}

std::ostream& operator<<(std::ostream &output,const PolygonVect &poly){

	output<< "<polygon points=\"";
	for(int i=0;i<poly.points.size();i++){
		output << poly.points[i].getX() << "," << poly.points[i].getY() << " ";
	}
	
	output<< "\" style=\"fill:" << poly.color << "\" />" << std::endl ;
	
	return output;

}


}

File: HW07/hw7/src/PolygonVect.java

import java.awt.Graphics;
import java.util.ArrayList;
import java.awt.Color;
/**
*This is the polygonvect class
*Extends from Polygon class 
*Uses ArrayList collection for holding points.
*@author Cihan Can Ayyıldız.
*/
public class PolygonVect extends Polygon{

	/**
	*Arraylist field of PolygonVect
	*This field holds polygonvect's points.
	*/
	private ArrayList<Point2D> points;

	/**
	*No parameter constructor for PolygonVect
	*/
	public PolygonVect(){
		points = new ArrayList<Point2D>();
		color = Color.RED;
	}

	/**
	*1 parameter constructor for PolygonVect
	*@param newpoints This is the new ArrayList object.It holds point2D
	*adding all objects of newpoints to end of our ArrayList.
	*/
	public PolygonVect(ArrayList<Point2D> newpoints){
		points = new ArrayList<Point2D>();
		points.addAll(newpoints);
		color = Color.RED;
	}

	/**
	*1 parameter constructor for PolygonVect
	*@param p new point2D object.
	*adding this point into our arraylist.
	*/
	public PolygonVect(Point2D p){
		points = new ArrayList<Point2D>();
		points.add(p);
		color = Color.RED;
	}

	/**
	*Conversion constructor for PolygonVect
	*@param circle this is an circle object.
	*Converting circle to polygonvect object.
	*/
	public PolygonVect(Circle circle){
		points = new ArrayList<Point2D>(100);
		for(int i=0;i<100;i++) {
			Point2D p = new Point2D();
	  		p.setX((circle.getCx() + circle.getRadius() * Math.cos(i * 2 * Math.PI / 100)));	// X-axis for polygon.
	  		p.setY((circle.getCy() + circle.getRadius() * Math.sin(i * 2 * Math.PI / 100)));	// Y-axis for polygon.
	  		points.add(p);
		}

		color = circle.getColor();
	}

	/**
	*Conversion constructor for PolygonVect
	*@param rect this is an rectangle object.
	*Converting rectangle to polygonvect object.
	*/
	public PolygonVect(Rectangle rect){
		points = new ArrayList<Point2D>();

		Point2D p = new Point2D();
		p.setX(rect.getX());
		p.setY(rect.getY());
		points.add(p);
		Point2D p1 = new Point2D();
		p1.setX(rect.getX());
		p1.setY(rect.getY()+rect.getHeight());
		points.add(p1);
		Point2D p2 = new Point2D();
		p2.setX(rect.getX()+rect.getWidth());
		p2.setY(rect.getY()+rect.getHeight());
		points.add(p2);
		Point2D p3 = new Point2D();
		p3.setX(rect.getX()+rect.getWidth());
		p3.setY(rect.getY());
		points.add(p3);
		
		color = rect.getColor();
	}

	/**
	*Conversion constructor for PolygonVect
	*@param tri this is an triangle object.
	*Converting  triangle to polygonvect object.
	*/
	public PolygonVect(Triangle tri){
		points = new ArrayList<Point2D>(3);

		Point2D p = new Point2D();
		p.setX(tri.get_x_axis_of_Point1());
		p.setY(tri.get_y_axis_of_Point1());
		points.add(p);
		Point2D p2 = new Point2D();
		p2.setX(tri.get_x_axis_of_Point2());
		p2.setY(tri.get_y_axis_of_Point2());
		points.add(p2);
		Point2D p3 = new Point2D();
		p3.setX(tri.get_x_axis_of_Point3());
		p3.setY(tri.get_y_axis_of_Point3());
		points.add(p3);

		color = tri.getColor();
	}

	/**
	*This is the Area method for PolygonVect
	*Overriding the area method.
	*Extends from Polygon class
	*@return area of polygonvect.
	*/
	public double Area(){
		double area = 0;

		for(int i=0 ; i<points.size()-1 ; i++){
			area += (points.get(i).getX() * points.get(i+1).getY() - points.get(i).getY() * points.get(i+1).getX() );
		}
		area += (points.get(points.size()-1).getX()*points.get(0).getY() - points.get(points.size()-1).getY() * points.get(0).getX() );
		area = Math.abs(area)/2.0;
	
		return area;
	}

	/**
	*This is the Perimeter method for PolygonVect
	*Overriding the perimeter method.
	*Extends from Polygon class
	*@return perimeter of polygonvect.
	*/
	public double Perimeter(){
		double perimeter = 0;

		for(int i=0 ; i<points.size()-1 ; i++){
			perimeter += Math.sqrt(Math.pow(points.get(i).getX()-points.get(i+1).getX(),2) + Math.pow(points.get(i).getY() - points.get(i+1).getY(),2));
		}
		perimeter += Math.sqrt(Math.pow(points.get(points.size()-1).getX()-points.get(0).getX(),2) + Math.pow(points.get(points.size()-1).getY() - points.get(0).getY(),2));

		return perimeter;
	}

	/**
	*Overriding compareTo method for comparing Areas.
	*@param s Shape object.
	*@return a negative integer, zero, or a positive integer as this object is less than, equal to, or greater than the specified object
	*/
	public int compareTo(Shape s){
		return Double.compare(Area(),s.Area());
	}


	/**
	*İncrementing position of PolygonVect 1.0
	*/

	public void increment(){
		Point2D p;
		for(int i=0 ; i<points.size() ; i++){
			p = points.get(i);
			p.setX(p.getX()+1);
			p.setY(p.getY()+1);
			points.set(i,p);
		}
	}

	/**
	*Decrementing position of PolygonVect 1.0
	*/
	public void decrement(){
		Point2D p;
		for(int i=0 ; i<points.size() ; i++){
			p = points.get(i);
			p.setX(p.getX()-1);
			p.setY(p.getY()-1);
			points.set(i,p);
		}
	}


	public void draw(Graphics g){
		
		int[] x_axis = new int[points.size()];
		for(int i=0;i<points.size();i++){
			x_axis[i] = (int)(points.get(i).getX());
		}
		int[] y_axis = new int[points.size()];
		for(int i=0;i<points.size();i++){
			y_axis[i] = (int)(points.get(i).getY());
		}

		g.setColor(getColor());
		g.fillPolygon( x_axis , y_axis , points.size());
		g.setColor(Color.BLACK);
		g.drawPolygon( x_axis , y_axis , points.size());
			
	}

}